### PR TITLE
Bugfix/unr 2891 rpc calls not traced

### PR DESCRIPTION
**Pull request by improbable-danny**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
- RPC traces are not continued properly 
- Some minor fixes
- BeginTrace was not behaving properly with the root span never being ended because it was never set active

#### Tests
Testing using the test gyms and observing results in stackdriver

#### Primary reviewers
@m-samiec @ImprobableVlad

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp [hunk 1]
```diff
@@ -432,8 +436,19 @@ bool USpatialLatencyTracer::ContinueLatencyTrace_Internal(const AActor* Actor, c
 		return false;
 	}
 
+	auto TraceTypeToStr = [](ETraceType Type) -> TCHAR*
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: I believe you should be able to use UClass::GetValueAsString here instead

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp [hunk 2]
```diff
@@ -586,7 +601,7 @@ USpatialLatencyTracer::TraceSpan* USpatialLatencyTracer::GetActiveTraceOrReadPay
 
 void USpatialLatencyTracer::WriteKeyFrameToTrace(const TraceSpan* Trace, const FString& TraceDesc)
 {
-	if (Trace != nullptr)
+	if (Trace != nullptr && TraceDesc.Len())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: Why this change?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp [hunk 3]
```diff
@@ -403,6 +403,10 @@ bool USpatialLatencyTracer::BeginLatencyTrace_Internal(const FString& TraceDesc,
 		OutLatencyPayload = FSpatialLatencyPayload(MoveTemp(TraceBytes), MoveTemp(SpanBytes));
 	}
 
+	TraceKey Key = GenerateNewTraceKey();
+	TraceMap.Add(Key, MoveTemp(NewTrace));
+	ActiveTraceKey = Key; // This will be followed by ContinueLatencyTrace
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: Shouldn't it be the responsibility of the user to pass the Trace back into the continue function? I feel we should move towards removing the `active` trace.

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: I think there was a bug in the initial version because the root span is never closed currently, ContinueLatencyTrace does not end the previous span which would mean the root never got closed - this caused some weird looking traces where the root was swapped with the first Continue trace :/ I'll investigate improving the situation instead

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: So I changed this code to just end the trace and then Continue will add it into the internal map, Sami has a doc for proposing better solutions for tracing (one of which includes removing the active trace concept)

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp (hunk 2 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: Removed now, probably not needed :D

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp [hunk 4]
```diff
@@ -449,10 +454,14 @@ bool USpatialLatencyTracer::ContinueLatencyTrace_Internal(const AActor* Actor, c
 	TraceMap.Add(Key, MoveTemp(TempSpan));
 	TraceMap.Remove(ActiveTraceKey);
 	ActiveTraceKey = InvalidTraceKey;
+	PayloadToTraceKeys.Remove(LatencyPayload);
+	PayloadToTraceKeys.Add(MakeTuple(OutLatencyPayloadContinue, Key)); // Add continued payload to tracking map. 
 
 	if (!GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking())
 	{
 		// We can't do any deeper tracing in the stack here so terminate these traces here
+		EndLatencyTrace(Key, TEXT("End of native tracing"));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: This is only the case if it's a property or rpc trace

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp [hunk 5]
```diff
@@ -566,6 +577,19 @@ USpatialLatencyTracer::TraceSpan* USpatialLatencyTracer::GetActiveTrace()
 
 USpatialLatencyTracer::TraceSpan* USpatialLatencyTracer::GetActiveTraceOrReadPayload(const FSpatialLatencyPayload& Payload)
 {
+	if (TraceKey* ExistingKey = PayloadToTraceKeys.Find(Payload)) // This occurs if the root was created on this machine 
+	{
+		USpatialLatencyTracer::TraceSpan* Span = TraceMap.Find(*ExistingKey);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: Move this inside the `if`

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyPayload.h [hunk 6]
```diff
@@ -24,4 +25,14 @@ struct SPATIALGDK_API FSpatialLatencyPayload
 
 	UPROPERTY()
 	TArray<uint8> SpanId;
+
+	// Required for TMap hash
+	bool operator== (const FSpatialLatencyPayload& Other) const
+	{
+		return TraceId == Other.TraceId && SpanId == Other.SpanId;
+	}
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyPayload.h]:
```suggestion
	}
	
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyPayload.h [hunk 7]
```diff
@@ -24,4 +25,14 @@ struct SPATIALGDK_API FSpatialLatencyPayload
 
 	UPROPERTY()
 	TArray<uint8> SpanId;
+
+	// Required for TMap hash
+	bool operator== (const FSpatialLatencyPayload& Other) const
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyPayload.h]:
```suggestion
	bool operator == (const FSpatialLatencyPayload& Other) const
```
